rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java

```diff
@@ -125,70 +127,107 @@ public static void undo(DataSourceProxy dataSourceProxy, String xid, long branch
-                if (rs != null) {
-                    rs.close();
+                conn = dataSourceProxy.getPlainConnection();
+
+                // The entire undo process should run in a local transaction.
+                conn.setAutoCommit(false);
+
+                // Find UNDO LOG
+                selectPST = conn.prepareStatement(SELECT_UNDO_LOG_SQL);
+                selectPST.setLong(1, branchId);
+                selectPST.setString(2, xid);
+                rs = selectPST.executeQuery();
+
+                boolean exists = false;
+                while (rs.next()) {
+                    exists = true;
+
+                    // It is possible that the server repeatedly sends a rollback request to roll back
+                    // the same branch transaction to multiple processes,
+                    // ensuring that only the undo_log in the normal state is processed.
+                    int state = rs.getInt("log_status");
+                    if (!canUndo(state)) {
+                        LOGGER.info("xid {} branch {}, ignore {} undo_log",
+                            xid, branchId, state);
+                        return;
+                    }
+
+                    Blob b = rs.getBlob("rollback_info");
+                    String rollbackInfo = StringUtils.blob2string(b);
+                    BranchUndoLog branchUndoLog = UndoLogParserFactory.getInstance().decode(rollbackInfo);
+
+                    for (SQLUndoLog sqlUndoLog : branchUndoLog.getSqlUndoLogs()) {
+                        TableMeta tableMeta = TableMetaCache.getTableMeta(dataSourceProxy, sqlUndoLog.getTableName());
+                        sqlUndoLog.setTableMeta(tableMeta);
+                        AbstractUndoExecutor undoExecutor = UndoExecutorFactory.getUndoExecutor(dataSourceProxy.getDbType(),
+                            sqlUndoLog);
+                        undoExecutor.executeOn(conn);
+                    }
                 }
-                if (selectPST != null) {
-                    selectPST.close();
+
+                // If undo_log exists, it means that the branch transaction has completed the first phase,
+                // we can directly roll back and clean the undo_log
+                // Otherwise, it indicates that there is an exception in the branch transaction,
+                // causing undo_log not to be written to the database.
+                // For example, the business processing timeout, the global transaction is the initiator rolls back.
+                // To ensure data consistency, we can insert an undo_log with GlobalFinished state
+                // to prevent the local transaction of the first phase of other programs from being correctly submitted.
+                // See https://github.com/alibaba/fescar/issues/489
+
+                if (exists) {
+                    deleteUndoLog(xid, branchId, conn);
+                    conn.commit();
+                    LOGGER.info("xid {} branch {}, undo_log deleted",
```

Comment:
logging placeholder seems is missed for state.

Reply:
Yes，good job!